File: packages/typescript-sdk/test/run-context.test.ts
import { describe, expect, test } from "bun:test";

import {
  createRunContext,
  type RunContextClient,
} from "../src/authoring/run-context";

const createMockClient = () => {
  const calls: { method: string; args: unknown }[] = [];

  const track = (method: string) => (input: unknown) => {
    calls.push({ method, args: input });
    if (method === "spawnRun") {
      return Promise.resolve({ id: "spawn_1" });
    }
    if (method === "continueRun") {
      return Promise.resolve({ id: "continue_1" });
    }
    if (method === "waitForEventRun") {
      return Promise.resolve({
        status: "waiting",
        eventKey: "test",
        triggerId: "t_1",
        expiresAt: "2026-01-01T00:00:00Z",
      });
    }
    return Promise.resolve({});
  };

  const client: RunContextClient = {
    checkpointRun: track("checkpointRun"),
    heartbeatRun: track("heartbeatRun"),
    progressRun: track("progressRun"),
    logRun: track("logRun"),
    usageRun: track("usageRun"),
    toolCallRun: track("toolCallRun"),
    outputRun: track("outputRun"),
    waitForEventRun: track("waitForEventRun"),
    spawnRun: track("spawnRun"),
    continueRun: track("continueRun"),
    annotateRun: track("annotateRun"),
    completeRun: track("completeRun"),
    failRun: track("failRun"),
    stateRun: track("stateRun"),
    getStateByRunId: track("getStateByRunId"),
    getStateByRunIdAndKey: track("getStateByRunIdAndKey"),
    deleteState: track("deleteState"),
    streamRun: track("streamRun"),
  };

  return { client, calls };
};

describe("createRunContext", () => {
  test("sets runId, attempt, and signal", () => {
    const { client } = createMockClient();
    const signal = AbortSignal.timeout(5000);
    const ctx = createRunContext(client, "run_123", { attempt: 3, signal });

    expect(ctx.runId).toBe("run_123");
    expect(ctx.attempt).toBe(3);
    expect(ctx.signal).toBe(signal);
  });

  test("defaults attempt to 1", () => {
    const { client } = createMockClient();
    const ctx = createRunContext(client, "run_123");
    expect(ctx.attempt).toBe(1);
  });

  test("checkpoint sends state with source sdk", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.checkpoint({ step: 5 });

    expect(calls).toHaveLength(1);
    expect(calls[0].method).toBe("checkpointRun");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { state: { step: 5 }, source: "sdk" },
    });
  });

  test("reportProgress sends percent and message", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.reportProgress(75, "Processing items");

    expect(calls[0].method).toBe("progressRun");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { percent: 75, message: "Processing items" },
    });
  });

  test("heartbeat sends correct pathParams", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.heartbeat();

    expect(calls[0].method).toBe("heartbeatRun");
    expect(calls[0].args).toEqual({ pathParams: { runID: "run_123" } });
  });

  test("logger methods are fire-and-forget", () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    // These should not return promises
    const infoResult = ctx.logger.info("hello", { key: "value" });
    const warnResult = ctx.logger.warn("warning");
    const errorResult = ctx.logger.error("error");

    expect(infoResult).toBeUndefined();
    expect(warnResult).toBeUndefined();
    expect(errorResult).toBeUndefined();

    // But they should have triggered calls
    expect(calls.length).toBeGreaterThanOrEqual(3);
    expect(calls[0].method).toBe("logRun");
  });

  test("reportUsage maps camelCase to snake_case", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.reportUsage?.({
      provider: "openai",
      model: "gpt-4o",
      promptTokens: 100,
      completionTokens: 50,
      totalTokens: 150,
      costMicrousd: 500,
    });

    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: {
        provider: "openai",
        model: "gpt-4o",
        prompt_tokens: 100,
        completion_tokens: 50,
        total_tokens: 150,
        cost_microusd: 500,
      },
    });
  });

  test("logToolCall maps camelCase to snake_case", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.logToolCall?.({
      toolName: "search",
      input: { query: "test" },
      output: { results: [] },
      durationMs: 150,
      status: "success",
    });

    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: {
        tool_name: "search",
        input: { query: "test" },
        output: { results: [] },
        duration_ms: 150,
        status: "success",
      },
    });
  });

  test("saveOutput sends key, value, and schema", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.saveOutput?.("result", { total: 42 }, { type: "object" });

    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { key: "result", value: { total: 42 }, schema: { type: "object" } },
    });
  });

  test("spawn sends job_slug not job_id", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    const result = await ctx.spawn?.({
      jobSlug: "my-job",
      projectId: "proj_1",
      payload: { key: "value" },
      priority: 5,
    });

    expect(result).toEqual({ id: "spawn_1" });
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: {
        job_slug: "my-job",
        project_id: "proj_1",
        payload: { key: "value" },
        priority: 5,
      },
    });
  });

  test("continue sends payload", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    const result = await ctx.continue?.({ next: "step" });

    expect(result).toEqual({ id: "continue_1" });
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { payload: { next: "step" } },
    });
  });

  test("waitForEvent maps to snake_case", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.waitForEvent?.("payment.completed", { timeoutSecs: 60 });

    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: {
        event_key: "payment.completed",
        timeout_secs: 60,
        notify_url: undefined,
      },
    });
  });

  test("annotate sends annotations", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.annotate?.({ team: "backend" });

    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { annotations: { team: "backend" } },
    });
  });

  test("complete and fail send correct bodies", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.complete?.({ output: "done" });
    await ctx.fail?.("something broke");

    expect(calls[0].method).toBe("completeRun");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { result: { output: "done" } },
    });
    expect(calls[1].method).toBe("failRun");
    expect(calls[1].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { error: "something broke" },
    });
  });

  test("state.set sends key and value via stateRun", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.state?.set("counter", 42);

    expect(calls[0].method).toBe("stateRun");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { key: "counter", value: 42 },
    });
  });

  test("state.get calls getStateByRunIdAndKey with key path param", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.state?.get("counter");

    expect(calls[0].method).toBe("getStateByRunIdAndKey");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123", key: "counter" },
    });
  });

  test("state.delete calls deleteState with key path param", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.state?.delete("counter");

    expect(calls[0].method).toBe("deleteState");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123", key: "counter" },
    });
  });

  test("state.list calls getStateByRunId", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.state?.list();

    expect(calls[0].method).toBe("getStateByRunId");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
    });
  });

  test("streamChunk calls streamRun with correct body", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_123");

    await ctx.streamChunk?.("Hello world", { streamId: "s1", done: false });

    expect(calls[0].method).toBe("streamRun");
    expect(calls[0].args).toEqual({
      pathParams: { runID: "run_123" },
      body: { chunk: "Hello world", stream_id: "s1", done: false },
    });
  });

  test("all methods pass correct runID in pathParams", async () => {
    const { client, calls } = createMockClient();
    const ctx = createRunContext(client, "run_abc");

    await ctx.checkpoint({ x: 1 });
    await ctx.heartbeat();
    await ctx.reportProgress(50);
    await ctx.reportUsage?.({ provider: "test", model: "test" });
    await ctx.logToolCall?.({ toolName: "test" });
    await ctx.saveOutput?.("k", { v: 1 });

    for (const call of calls) {
      expect(
        (call.args as { pathParams: { runID: string } }).pathParams.runID
      ).toBe("run_abc");
    }
  });
});

File: packages/go-sdk/authoring/run_context_test.go
package authoring

import (
	"context"
	"testing"
)

func TestCreateRunContext_DefaultFields(t *testing.T) {
	ctx, _ := CreateTestContext("run-1")

	if ctx.RunID != "run-1" {
		t.Errorf("expected RunID 'run-1', got %q", ctx.RunID)
	}
	if ctx.Attempt != 1 {
		t.Errorf("expected Attempt 1, got %d", ctx.Attempt)
	}
	if ctx.Logger == nil {
		t.Error("expected Logger to be non-nil")
	}
}

func TestCreateRunContext_WithAttempt(t *testing.T) {
	ctx, _ := CreateTestContext("run-2", WithAttempt(3))

	if ctx.Attempt != 3 {
		t.Errorf("expected Attempt 3, got %d", ctx.Attempt)
	}
}

func TestCreateRunContext_WithContext(t *testing.T) {
	bg := context.Background()
	child, cancel := context.WithCancel(bg)
	defer cancel()

	ctx, _ := CreateTestContext("run-3", WithContext(child))

	if ctx.Ctx != child {
		t.Error("expected Ctx to match provided context")
	}
}

func TestRunContext_Checkpoint(t *testing.T) {
	ctx, record := CreateTestContext("run-cp")

	err := ctx.Checkpoint(map[string]any{"step": 1})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(record.Checkpoints) != 1 {
		t.Fatalf("expected 1 checkpoint, got %d", len(record.Checkpoints))
	}
	if record.Checkpoints[0]["step"] != 1 {
		t.Error("expected checkpoint state with step=1")
	}
}

func TestRunContext_ReportProgress(t *testing.T) {
	t.Run("without message", func(t *testing.T) {
		ctx, record := CreateTestContext("run-prog")
		err := ctx.ReportProgress(0.5)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(record.ProgressUpdates) != 1 {
			t.Fatalf("expected 1 progress update, got %d", len(record.ProgressUpdates))
		}
		if record.ProgressUpdates[0]["percent"] != 0.5 {
			t.Error("expected percent 0.5")
		}
		if _, ok := record.ProgressUpdates[0]["message"]; ok {
			t.Error("expected no message key")
		}
	})

	t.Run("with message", func(t *testing.T) {
		ctx, record := CreateTestContext("run-prog2")
		err := ctx.ReportProgress(0.75, "almost done")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if record.ProgressUpdates[0]["message"] != "almost done" {
			t.Error("expected message 'almost done'")
		}
	})
}

func TestRunContext_Heartbeat(t *testing.T) {
	ctx, record := CreateTestContext("run-hb")

	_ = ctx.Heartbeat()
	_ = ctx.Heartbeat()

	if record.Heartbeats != 2 {
		t.Errorf("expected 2 heartbeats, got %d", record.Heartbeats)
	}
}

func TestRunContext_ReportUsage(t *testing.T) {
	ctx, record := CreateTestContext("run-usage")
	prompt := 100
	completion := 50
	total := 150
	cost := 500

	err := ctx.ReportUsage(UsageReport{
		Provider:         "openai",
		Model:            "gpt-4",
		PromptTokens:     &prompt,
		CompletionTokens: &completion,
		TotalTokens:      &total,
		CostMicrousd:     &cost,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(record.UsageReports) != 1 {
		t.Fatalf("expected 1 usage report, got %d", len(record.UsageReports))
	}
	report := record.UsageReports[0]
	if report["provider"] != "openai" {
		t.Error("expected provider openai")
	}
	if report["model"] != "gpt-4" {
		t.Error("expected model gpt-4")
	}
	if report["prompt_tokens"] != 100 {
		t.Error("expected prompt_tokens 100")
	}
	if report["cost_microusd"] != 500 {
		t.Error("expected cost_microusd 500")
	}
}

func TestRunContext_LogToolCall(t *testing.T) {
	ctx, record := CreateTestContext("run-tc")
	dur := 42

	err := ctx.LogToolCall(ToolCallReport{
		ToolName:   "search",
		Input:      map[string]any{"query": "test"},
		Output:     map[string]any{"results": 5},
		DurationMs: &dur,
		Status:     "success",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(record.ToolCalls) != 1 {
		t.Fatalf("expected 1 tool call, got %d", len(record.ToolCalls))
	}
	tc := record.ToolCalls[0]
	if tc["tool_name"] != "search" {
		t.Error("expected tool_name search")
	}
	if tc["status"] != "success" {
		t.Error("expected status success")
	}
}

func TestRunContext_SaveOutput(t *testing.T) {
	t.Run("without schema", func(t *testing.T) {
		ctx, record := CreateTestContext("run-out")
		err := ctx.SaveOutput("result", map[string]any{"value": 42})
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(record.Outputs) != 1 {
			t.Fatalf("expected 1 output, got %d", len(record.Outputs))
		}
		if record.Outputs[0]["key"] != "result" {
			t.Error("expected key 'result'")
		}
	})

	t.Run("with schema", func(t *testing.T) {
		ctx, record := CreateTestContext("run-out2")
		schema := map[string]any{"type": "object"}
		err := ctx.SaveOutput("data", map[string]any{"v": 1}, schema)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if record.Outputs[0]["schema"] == nil {
			t.Error("expected schema to be set")
		}
	})
}

func TestRunContext_State(t *testing.T) {
	ctx, record := CreateTestContext("run-state")

	t.Run("set and get", func(t *testing.T) {
		err := ctx.State.Set("foo", "bar")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if record.StateStore["foo"] != "bar" {
			t.Error("expected state foo=bar")
		}

		result, err := ctx.State.Get("foo")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		m, ok := result.(map[string]any)
		if !ok {
			t.Fatalf("expected map result, got %T", result)
		}
		if m["value"] != "bar" {
			t.Error("expected value bar")
		}
	})

	t.Run("delete", func(t *testing.T) {
		_ = ctx.State.Set("to_delete", "val")
		err := ctx.State.Delete("to_delete")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if _, ok := record.StateStore["to_delete"]; ok {
			t.Error("expected key to be deleted")
		}
	})

	t.Run("list", func(t *testing.T) {
		_ = ctx.State.Set("a", 1)
		_ = ctx.State.Set("b", 2)
		items, err := ctx.State.List()
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(items) < 2 {
			t.Errorf("expected at least 2 items, got %d", len(items))
		}
	})
}

func TestRunContext_StreamChunk(t *testing.T) {
	t.Run("simple chunk", func(t *testing.T) {
		ctx, record := CreateTestContext("run-stream")
		err := ctx.StreamChunk("hello")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(record.StreamChunks) != 1 {
			t.Fatalf("expected 1 chunk, got %d", len(record.StreamChunks))
		}
		if record.StreamChunks[0]["chunk"] != "hello" {
			t.Error("expected chunk 'hello'")
		}
	})

	t.Run("with options", func(t *testing.T) {
		ctx, record := CreateTestContext("run-stream2")
		err := ctx.StreamChunk("world", StreamChunkOption{StreamID: "s1", Done: true})
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if record.StreamChunks[0]["stream_id"] != "s1" {
			t.Error("expected stream_id 's1'")
		}
		if record.StreamChunks[0]["done"] != true {
			t.Error("expected done=true")
		}
	})
}

func TestRunContext_WaitForEvent(t *testing.T) {
	ctx, record := CreateTestContext("run-event")

	result, err := ctx.WaitForEvent("payment.completed")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["received"] != true {
		t.Error("expected received=true")
	}
	if len(record.Events) != 1 {
		t.Fatalf("expected 1 event, got %d", len(record.Events))
	}
	if record.Events[0]["event_key"] != "payment.completed" {
		t.Error("expected event_key 'payment.completed'")
	}
}

func TestRunContext_WaitForEvent_WithOptions(t *testing.T) {
	ctx, record := CreateTestContext("run-event-opts")
	timeout := 30

	_, err := ctx.WaitForEvent("approval", WaitForEventOption{TimeoutSecs: &timeout, NotifyUrl: "https://example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if record.Events[0]["timeout_secs"] != 30 {
		t.Error("expected timeout_secs 30")
	}
	if record.Events[0]["notify_url"] != "https://example.com" {
		t.Error("expected notify_url")
	}
}

func TestRunContext_Spawn(t *testing.T) {
	ctx, record := CreateTestContext("run-spawn")
	priority := 5

	result, err := ctx.Spawn(SpawnOptions{
		JobSlug:   "child-job",
		ProjectID: "proj-1",
		Payload:   map[string]any{"input": "data"},
		Priority:  &priority,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["run_id"] != "spawned-123" {
		t.Error("expected run_id 'spawned-123'")
	}
	if len(record.Spawns) != 1 {
		t.Fatalf("expected 1 spawn, got %d", len(record.Spawns))
	}
	if record.Spawns[0]["job_slug"] != "child-job" {
		t.Error("expected job_slug 'child-job'")
	}
	if record.Spawns[0]["priority"] != 5 {
		t.Error("expected priority 5")
	}
}

func TestRunContext_Continue(t *testing.T) {
	t.Run("without payload", func(t *testing.T) {
		ctx, _ := CreateTestContext("run-cont")
		result, err := ctx.Continue()
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result["continued"] != true {
			t.Error("expected continued=true")
		}
	})

	t.Run("with payload", func(t *testing.T) {
		ctx, record := CreateTestContext("run-cont2")
		_, err := ctx.Continue(map[string]any{"next": "step"})
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(record.Continuations) != 1 {
			t.Fatalf("expected 1 continuation, got %d", len(record.Continuations))
		}
	})
}

func TestRunContext_Annotate(t *testing.T) {
	ctx, record := CreateTestContext("run-ann")

	err := ctx.Annotate(map[string]string{"env": "prod", "team": "backend"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(record.Annotations) != 1 {
		t.Fatalf("expected 1 annotation, got %d", len(record.Annotations))
	}
	if record.Annotations[0]["env"] != "prod" {
		t.Error("expected annotation env=prod")
	}
}

func TestRunContext_Complete(t *testing.T) {
	t.Run("without result", func(t *testing.T) {
		ctx, record := CreateTestContext("run-comp")
		err := ctx.Complete()
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !record.Completed {
			t.Error("expected Completed=true")
		}
	})

	t.Run("with result", func(t *testing.T) {
		ctx, record := CreateTestContext("run-comp2")
		err := ctx.Complete(map[string]any{"output": "done"})
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !record.Completed {
			t.Error("expected Completed=true")
		}
		if record.Result["output"] != "done" {
			t.Error("expected result output=done")
		}
	})
}

func TestRunContext_Fail(t *testing.T) {
	ctx, record := CreateTestContext("run-fail")

	err := ctx.Fail("something went wrong")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !record.Failed {
		t.Error("expected Failed=true")
	}
	if record.FailError != "something went wrong" {
		t.Errorf("expected FailError 'something went wrong', got %q", record.FailError)
	}
}
